### PR TITLE
UNOMI-562

### PR DESCRIPTION
**Pull request by dgriffon**
UNOMI-562
- add json schema for ContextRequest
- Use custom deserializer for ContextRequest and EventCollectorRequest to perform json schema validation

### CODE REVIEW

File: api/src/main/java/org/apache/unomi/api/services/SchemaRegistry.java [hunk 1]
```diff
@@ -51,7 +52,7 @@ public interface SchemaRegistry {
      * @param schemaId id of the schema used for the validation
      * @return true is the object is valid
      */
-    boolean isValid(Object object, String schemaId);
+    boolean isValid(JsonNode object, String schemaId);
```
Comment (sergehuber) [inline api/src/main/java/org/apache/unomi/api/services/SchemaRegistry.java:55]: Could we maybe rename the object to node and change the Javadoc ?

File: rest/src/main/java/org/apache/unomi/rest/deserializers/ContextRequestDeserializer.java [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+import javax.ws.rs.BadRequestException;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Custom deserializer for ContextRequest that do validate the object using JSon Schema
```
Comment (sergehuber) [inline rest/src/main/java/org/apache/unomi/rest/deserializers/ContextRequestDeserializer.java:40]: Minor typo: should be JSON schema

File: rest/src/main/java/org/apache/unomi/rest/deserializers/ContextRequestDeserializer.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+                }
+                cr.setEvents(filteredEvents);
+            }
+            final JsonNode filtersNode = node.get("filters");
+            if (filtersNode instanceof ArrayNode) {
+                ArrayNode filters = (ArrayNode) filtersNode;
+                List<PersonalizationService.PersonalizedContent> f = new ArrayList<>();
+                filters.elements().forEachRemaining(el -> {
+                    try {
+                        jsonParser.getCodec().treeToValue(el, PersonalizationService.PersonalizedContent.class);
```
Comment (sergehuber) [inline rest/src/main/java/org/apache/unomi/rest/deserializers/ContextRequestDeserializer.java:96]: Are we not missing f.add here ?

File: rest/src/main/java/org/apache/unomi/rest/deserializers/ContextRequestDeserializer.java [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+                });
+                cr.setFilters(f);
+            }
+            final JsonNode personalizationsNode = node.get("personalizations");
+            if (personalizationsNode instanceof ArrayNode) {
+                ArrayNode personalizations = (ArrayNode) personalizationsNode;
+                List<PersonalizationService.PersonalizationRequest> p = new ArrayList<>();
+                personalizations.elements().forEachRemaining(el -> {
+                    try {
+                        jsonParser.getCodec().treeToValue(el, PersonalizationService.PersonalizationRequest.class);
```
Comment (sergehuber) [inline rest/src/main/java/org/apache/unomi/rest/deserializers/ContextRequestDeserializer.java:109]: Same comment here ? p.add is missing ?

File: rest/src/main/java/org/apache/unomi/rest/deserializers/ContextRequestDeserializer.java [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+                        // Unable to deserialize, ignore the entry
+                    }
+                });
+                cr.setPersonalizations(p);
+            }
+            if (node.get("profileOverrides") != null) {
+                cr.setProfileOverrides(jsonParser.getCodec().treeToValue(node.get("profileOverrides"), Profile.class));
+            }
+            if (node.get("sessionPropertiesOverrides") != null) {
+                jsonParser.getCodec().treeToValue(node.get("sessionPropertiesOverrides"), Map.class);
```
Comment (sergehuber) [inline rest/src/main/java/org/apache/unomi/rest/deserializers/ContextRequestDeserializer.java:120]: Also here it seems we are not assigning the value

File: services/src/main/java/org/apache/unomi/services/impl/events/EventServiceImpl.java [hunk 6]
```diff
@@ -139,7 +141,8 @@ public boolean isEventAllowed(Event event, String thirdPartyId) {
     }
 
     public boolean isEventValid(Event event) {
-        return schemaRegistry.isValid(event, "https://unomi.apache.org/schemas/json/events/" + event.getEventType() + "/1-0-0");
+
```
Comment (sergehuber) [inline services/src/main/java/org/apache/unomi/services/impl/events/EventServiceImpl.java:144]: Does this method still make sense to keep if all validation was done at the endpoint level ?

File: services/src/main/resources/META-INF/cxs/schemas/contextrequest.json [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    "filters": {
+      "type": ["null", "array"],
+      "items": {
+        "$ref": "https://unomi.apache.org/schemas/json/personalization/personalizedcontent/1-0-0"
+      }
+    },
+    "personalizations": {
+      "type": ["null", "array"],
+      "items": {
+        "$ref": "https://unomi.apache.org/schemas/json/event/1-0-0"
```
Comment (sergehuber) [inline services/src/main/resources/META-INF/cxs/schemas/contextrequest.json:43]: This looks wrong, shouldn't it be personalizedRequest instead ?

File: services/src/main/resources/META-INF/cxs/schemas/contextrequest.json [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+      }
+    },
+    "profileOverrides": {
+      "$ref": "https://unomi.apache.org/schemas/json/profile/1-0-0"
+    },
+    "sessionPropertiesOverrides": {
+      "type": ["null", "object"]
```
Comment (sergehuber) [inline services/src/main/resources/META-INF/cxs/schemas/contextrequest.json:50]: We might need to set a maximum here like : maxProperties : 50

**Review by sergehuber — CHANGES_REQUESTED** [review]
Hi David, 
Thanks for your PR. I'm not sure if it was complete but since it was not in draft I assumed it was. I found a few issues with it and put some comments in the code. Please don't hesitate to reach me if needed.
Regards,
  Serge...

File: rest/src/main/java/org/apache/unomi/rest/deserializers/ContextRequestDeserializer.java (hunk 4 above)
Comment (dgriffon) [reply rest/src/main/java/org/apache/unomi/rest/deserializers/ContextRequestDeserializer.java]: done

File: rest/src/main/java/org/apache/unomi/rest/deserializers/ContextRequestDeserializer.java (hunk 3 above)
Comment (dgriffon) [reply rest/src/main/java/org/apache/unomi/rest/deserializers/ContextRequestDeserializer.java]: done

File: rest/src/main/java/org/apache/unomi/rest/deserializers/ContextRequestDeserializer.java (hunk 5 above)
Comment (dgriffon) [reply rest/src/main/java/org/apache/unomi/rest/deserializers/ContextRequestDeserializer.java]: done

File: services/src/main/java/org/apache/unomi/services/impl/events/EventServiceImpl.java (hunk 6 above)
Comment (dgriffon) [reply services/src/main/java/org/apache/unomi/services/impl/events/EventServiceImpl.java]: indeed, removed as suggested

**Review by sergehuber — APPROVED** [review]

**Pull request merged**